controllers/idle_handlers.go

```diff
@@ -18,103 +18,208 @@ package controllers
-		_, err = r.Executor.Execute("unshare", "-m", "/bin/sh", "-c",
-			fmt.Sprintf("\"mount -o remount,rw /sysroot && rm -rf %s\"", sysrootPath))
-		if err != nil {
-			return fmt.Errorf("removing stateroot failed: %w", err)
+	failures := 0
+	for _, stateroot := range undeployStateroots {
+		if stateroot == bootedStateroot {
+			continue
+		}
+		if err = removeStateroot(r.Executor, stateroot); err != nil {
+			r.Log.Error(err, "failed to remove stateroot", "stateroot", stateroot)
+			failures += 1
 		}
 	}
 
-	return nil
-}
-
-//nolint:unparam
-func (r *ImageBasedUpgradeReconciler) handleAbort(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
-	// Cleanup precaching resources
-	r.Log.Info("Cleanup precaching resources")
-	if err := r.Precache.Cleanup(ctx); err != nil {
-		r.Log.Info("Failed to cleanup precaching resources")
-		return requeueWithError(err)
+	if failures == 0 {
+		return nil
 	}
+	return fmt.Errorf("failed to remove %d stateroots", failures)
+}
 
-	stateroot := getDesiredStaterootName(ibu)
-	r.Log.Info("Cleanup stateroot", "stateroot", stateroot)
-	err := r.cleanupUnbootedStateroot(stateroot)
+func (r *ImageBasedUpgradeReconciler) cleanupUnbootedStateroot(stateroot string) error {
+	status, err := r.RPMOstreeClient.QueryStatus()
 	if err != nil {
-		return requeueWithError(err)
+		return err
 	}
-	err = cleanupIBUFiles()
-	return requeueWithError(err)
-}
 
-//nolint:unparam
-func (r *ImageBasedUpgradeReconciler) handleAbortFailure(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
-
-	// TODO actual steps
-	// If succeeds, return doNotRequeue
-	return doNotRequeue(), nil
-}
-
-//nolint:unparam
-func (r *ImageBasedUpgradeReconciler) handleFinalize(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	undeployIndices := make([]int, 0)
+	for index, deployment := range status.Deployments {
+		if deployment.OSName != stateroot {
+			continue
+		}
+		// make sure none of the deployments in the stateroot is booted
+		if deployment.Booted {
+			return fmt.Errorf("failed abort: deployment %d in stateroot %s is booted", index, stateroot)
+		}
+		undeployIndices = append(undeployIndices, index)
+	}
+	// since undeploy shifts the order, undeploy in the reverse order
+	for i := len(undeployIndices) - 1; i >= 0; i-- {
+		_, err = r.Executor.Execute("ostree", "admin", "undeploy", fmt.Sprint(undeployIndices[i]))
+		if err != nil {
+			return fmt.Errorf("ostree undeploy %d failed: %w", i, err)
+		}
+	}
 
-	// TODO actual steps
-	// If succeeds, return doNotRequeue
-	return doNotRequeue(), nil
+	return removeStateroot(r.Executor, stateroot)
 }
 
-//nolint:unparam
-func (r *ImageBasedUpgradeReconciler) handleFinalizeFailure(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
-
-	// TODO actual steps
-	// If succeeds, return doNotRequeue
-	return doNotRequeue(), nil
-}
+func removeStateroot(executor ops.Execute, stateroot string) error {
```

Comment:
Having this piece in a separate function concerns me a little, due to the risk of someone calling it without having undeployed all the deployments in that stateroot. If there are multiple places where we may be wiping the stateroot, include the undeploy commands here. If not, if only cleanupUnbootedStateroot would be called, then have this step inline in that function

Reply:
There are multiple places, and can't undeploy here. because all the deployments of the stateroot should be undeployed before removing stateroot files.
I will copy it back into the two functions.

Reply:
But you could have a function that wipes a specified stateroot like:
```
func removeStateroot(args, stateroot) {
    // 1. Get the list of deployments in specified stateroot
    // 2. Undeploy the deployments
    // 3. Wipe out the stateroot
}
```

and have that function called from elsewhere. Or you can have a single removeUnbootedStateroots that also determines the stateroot to be deleted, like you're doing here... because we're never going to be removing the booted stateroot, after all.

But it's unsafe to delete the stateroot without first undeploying. So either they're in the same function, or you do a check before deleting the stateroot to ensure it is clean. My preference would be to have them in the same function.

Reply:
this is what `cleanupUnbootedStateroot(stateroot)` does.
There is also `cleanupUnbootedStateroots()` that does all this for all stateroots.